File: 1140. Stone Game II.java
class Solution {
    public int stoneGameII(int[] piles) {
        int n = piles.length;
        int[] suffixSums = new int[n + 1];
        int[][] memo = new int[n][n + 1];

        for (int i = n - 1; i >= 0; i--) {
            suffixSums[i] = suffixSums[i + 1] + piles[i];
        }

        
        return getMaxStones(1, 0, suffixSums, memo);
    }

    private int getMaxStones(int M, int currentPile, int[] suffixSums, int[][] memo) {
        int n = suffixSums.length - 1;

       
        if (currentPile >= n) return 0;        
        if (memo[currentPile][M] != 0) return memo[currentPile][M];
        if (currentPile + 2 * M >= n) {
            return suffixSums[currentPile];
        }

        int maxStones = 0;

        for (int x = 1; x <= 2 * M; x++) {
            int nextPile = currentPile + x;
            int currentStones = suffixSums[currentPile] - getMaxStones(Math.max(M, x), nextPile, suffixSums, memo);
            maxStones = Math.max(maxStones, currentStones);
        }

        memo[currentPile][M] = maxStones;
        return maxStones;
    }
}

File: 1140. Stone Game II.rs
impl Solution {
    pub fn stone_game_ii(piles: Vec<i32>) -> i32 {
        let n = piles.len();
        let stride = n + 1;
        let mut suffix_sums: Vec<i32> = Vec::new();
        let mut k: usize = 0;
        while k < stride {
            suffix_sums.push(0i32);
            k += 1;
        }
        let mut si = n;
        while si > 0 {
            si -= 1;
            suffix_sums[si] = suffix_sums[si + 1] + piles[si];
        }
        let total = stride * stride;
        let mut dp: Vec<i32> = Vec::new();
        let mut k2: usize = 0;
        while k2 < total {
            dp.push(0i32);
            k2 += 1;
        }
        let mut i: usize = n;
        while i > 0 {
            i -= 1;
            let mut m: usize = 1;
            while m <= n {
                if i + 2 * m >= n {
                    dp[i * stride + m] = suffix_sums[i];
                } else {
                    let mut best: i32 = suffix_sums[i] - dp[(i + 1) * stride + m];
                    let mut x: usize = 2;
                    while x <= 2 * m {
                        let new_m: usize = if x > m { x } else { m };
                        let score: i32 = suffix_sums[i] - dp[(i + x) * stride + new_m];
                        if score > best {
                            best = score;
                        }
                        x += 1;
                    }
                    dp[i * stride + m] = best;
                }
                m += 1;
            }
        }
        dp[1]
    }
}
